src/background/services/balances/handlers/getTotalBalanceForWallet/getTotalBalanceForWallet.ts

```diff
@@ -0,0 +1,160 @@
+    private accountsService: AccountsService,
+    private balanceAggregatorService: BalanceAggregatorService
+  ) {}
+
+  #getAddressesActivity = (addresses: string[]) =>
+    this.glacierService.getChainIdsForAddresses({
+      addresses,
+      network: this.networkService.isMainnet() ? Network.MAINNET : Network.FUJI,
+    });
+
+  async #findUnderivedAccounts(walletId: string, derivedAccounts: Account[]) {
+    const secrets = await this.secretService.getWalletAccountsSecretsById(
+      walletId
+    );
+
+    const derivedWalletAddresses = getAllAddressesForAccounts(
+      derivedAccounts ?? []
+    );
+    const derivedAddressesUnprefixed = derivedWalletAddresses.map((addr) =>
+      addr.replace(/^[PXC]-/i, '')
+    );
+    const underivedXPChainAddresses = secrets.xpubXP
+      ? (
+          await getAccountsWithActivity(
+            secrets.xpubXP,
+            await this.networkService.getAvalanceProviderXP(),
+            this.#getAddressesActivity
+          )
+        ).filter((address) => !derivedAddressesUnprefixed.includes(address))
+      : [];
+
+    return underivedXPChainAddresses.map<Partial<Account>>((address) => ({
+      addressPVM: `P-${address}`,
+      addressAVM: `X-${address}`,
+    }));
+  }
+
+  handle: HandlerType['handle'] = async ({ request }) => {
+    const { walletId } = request.params;
+    const requestsImportedAccounts = isImportedAccountsRequest(walletId);
+
+    try {
+      const allAccounts = this.accountsService.getAccounts();
+      const derivedAccounts = requestsImportedAccounts
+        ? Object.values(allAccounts.imported ?? {})
+        : allAccounts.primary[walletId] ?? [];
+
+      if (!derivedAccounts.length) {
+        return {
+          ...request,
+          result: {
+            totalBalanceInCurrency: 0,
+            hasBalanceOnUnderivedAccounts: false,
+          },
+        };
+      }
+
+      const underivedAccounts = requestsImportedAccounts
+        ? []
+        : await this.#findUnderivedAccounts(walletId, derivedAccounts);
+
+      const networksIncludedInTotal = getIncludedNetworks(
+        this.networkService.isMainnet(),
+        await this.networkService.activeNetworks.promisify(),
+        await this.networkService.getFavoriteNetworks()
+      );
+
+      // Get balance for derived addresses
+      const { tokens: derivedAddressesBalances } =
+        await this.balanceAggregatorService.getBalancesForNetworks(
+          networksIncludedInTotal,
+          derivedAccounts,
+          [TokenType.NATIVE, TokenType.ERC20]
+        );
+
+      let totalBalanceInCurrency = calculateTotalBalanceForAccounts(
+        derivedAddressesBalances,
+        derivedAccounts,
+        networksIncludedInTotal
+      );
+      let hasBalanceOnUnderivedAccounts = false;
+
+      if (underivedAccounts.length > 0) {
+        // Get balance for underived addresses for X- and P-Chain.
+        // We DO NOT cache this response. When fetching balances for multiple X/P addresses at once,
```

Comment:
Should we add a follow up to cache this response? Seems like it can take quite long for people with many UTXOs

Reply:
Ticket: https://ava-labs.atlassian.net/browse/CP-9628